src/metabase/pulse/markdown.clj

```diff
@@ -0,0 +1,320 @@
+      (str "*" joined-content "*")
+
+      :italic
+      (str "_" joined-content "_")
+
+      :code
+      (str "`" joined-content "`")
+
+      :codeblock
+      (str "```\n" joined-content "```")
+
+      :blockquote
+      (let [lines (str/split-lines joined-content)]
+        (str/join "\n" (map #(str ">" %) lines)))
+
+      :link
+      (let [resolved-uri (resolve-uri (:href attrs))]
+        (if (contains? #{:image :image-ref} (:tag (first content)))
+          ;; If this is a linked image, add link target on separate line after image placeholder
+          (str joined-content "\n(" resolved-uri ")")
+          (str "<" resolved-uri "|" joined-content ">")))
+
+      :link-ref
+      (if-let [resolved-uri (resolve-uri (-> attrs :reference :attrs :url))]
+        (str "<" resolved-uri "|" joined-content ">")
+        joined-content)
+
+      :auto-link
+      (str "<" (:href attrs) ">")
+
+      :mail-link
+      (str "<" (:address attrs) ">")
+
+      ;; list items might have nested lists or other elements, which should have their indentation level increased
+      (:unordered-list-item :ordered-list-item)
+      (let [indented-content (->> (rest resolved-content)
+                                  str/join
+                                  str/split-lines
+                                  (map #(str "    " %))
+                                  (str/join "\n"))]
+        (if-not (str/blank? indented-content)
+          (str (first resolved-content) indented-content "\n")
+          joined-content))
+
+      :unordered-list
+      (str/join (map #(str "• " %) resolved-content))
+
+      :ordered-list
+      (str/join (map-indexed #(str (inc %1) ". " %2) resolved-content))
+
+      ;; Replace images with text that links to source, including alt text if available
+      :image
+      (let [{:keys [src alt]} attrs]
+        (if (str/blank? alt)
+          (str "<" src "|[Image]>")
+          (str "<" src "|[Image: " alt "]>")))
+
+      :image-ref
+      (let [src (-> attrs :reference :attrs :url)
+            alt joined-content]
+        (if (str/blank? alt)
+          (str "<" src "|[Image]>")
+          (str "<" src "|[Image: " alt "]>")))
+
+      :html-entity
+      (some->> content
+               (get @html-entities)
+               (:characters))
+
+      joined-content)))
+
+(def ^:private ^:dynamic *slack-mrkdwn-length-limit*
+  3000)
+
+(defn- truncate-mrkdwn
+  "If a string is greater than Slack's length limit, truncates it to fit the limit and
+  adds an ellipsis character to the end."
+  [mrkdwn]
+  (if (> (count mrkdwn) *slack-mrkdwn-length-limit*)
+    (-> mrkdwn
+        (subs 0 (dec *slack-mrkdwn-length-limit*))
+        (str "…"))
+    mrkdwn))
+
+(defmulti process-markdown
+  "Converts a markdown string from a virtual card into a form that can be sent to the provided channel type
+  (mrkdwn for Slack; HTML for email)."
+  (fn [_markdown channel-type] channel-type))
+
+(defmethod process-markdown :slack
```

Comment:
I think the key here should be `:mrkdwn`. We have `:slack` and its compatriot `:email`, but i think that distinction doesn't need to bleed in here. It seems like given markdown, we can a) leave it as markdown, b) translate it into a similar language `mrkdwn`, or translate it into html, and that's the distinction in this part of the code.

Reply:
That's fair, I'll make this change, although I could see the logic both ways. In the future if we add support for another platform, it might have a different formatting syntax but not have a name for it. E.g. we add support for Discord, or whatever, and have to add a `:discord` method if there's nothing else to call it.

Reply:
true, but wouldn't that be support for the markdown flavor that discord supports and not necessarily tied to just discord? I guess i like the distinction that slack supports `:mrkdwn` and we can translate that. Discord supports `:dscrd` or whatever and those are just simple languages. And other platforms might support html and it would be weird to reuse `:email` for that.